infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -175,231 +186,272 @@ public void handleTailRequest(CorfuPayloadMsg<TailsRequest> msg, ChannelHandlerC
+            r.sendResponse(getResponseMsg(responseHeader, getWrongEpochErrorMsg(wee.getCorrectEpoch())), ctx);
         } else if (ex.getCause() instanceof OverwriteException) {
             OverwriteException owe = (OverwriteException) ex.getCause();
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE
-                    .payloadMsg(owe.getOverWriteCause().getId()));
-        } else if (ex.getCause() instanceof DataOutrankedException) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_OUTRANKED.msg());
-        } else if (ex.getCause() instanceof ValueAdoptedException) {
+            responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            r.sendResponse(getResponseMsg(responseHeader, getOverwriteErrorMsg(owe.getOverWriteCause().getId())), ctx);
+        } else if (ex.getCause() instanceof  DataOutrankedException) {
+            responseHeader = getHeaderMsg(req.getHeader(), false, false);
+            r.sendResponse(getResponseMsg(responseHeader, getDataOutrankedErrorMsg()), ctx);
+        } else if (ex.getCause() instanceof  ValueAdoptedException) {
             ValueAdoptedException vae = (ValueAdoptedException) ex.getCause();
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_VALUE_ADOPTED.payloadMsg(vae.getReadResponse()));
+            responseHeader = getHeaderMsg(req.getHeader(), false, false);
+            r.sendResponse(getResponseMsg(responseHeader, getValueAdoptedErrorMsg(vae.getReadResponse())), ctx);
         } else if (ex.getCause() instanceof TrimmedException) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_TRIMMED.msg());
+            responseHeader = getHeaderMsg(req.getHeader(), false, false);
+            r.sendResponse(getResponseMsg(responseHeader, getTrimmedErrorMsg()), ctx);
         } else {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_SERVER_EXCEPTION.payloadMsg(new ExceptionMsg(ex)));
+            responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            r.sendResponse(getResponseMsg(responseHeader, getUnknownErrorMsg(ex)), ctx);
             throw new LogUnitException(ex);
         }
     }
 
     /**
      * Service an incoming write request.
      */
-    @ServerHandler(type = CorfuMsgType.WRITE)
-    public void write(CorfuPayloadMsg<WriteRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
-        LogData logData = (LogData) msg.getPayload().getData();
-        log.debug("log write: type: {}, address: {}, streams: {}", logData.getType(),
-                logData.getToken(), logData.getBackpointerMap());
+    @RequestHandler(type = PayloadCase.WRITE_LOG_REQUEST)
+    public void handleWrite(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        LogData logData = getLogData(req.getPayload().getWriteLogRequest().getLogData());
+        log.debug("handleWrite: type: {}, address: {}, streams: {}",
+                logData.getType(), logData.getToken(), logData.getBackpointerMap());
```

Comment:
Is it better to name it `LogWriteRequest` vs `WriteLogRequest?` . Not sure if there is a convention about verb/noun vs noun/verb